benchmarking/benchmark.py

```diff
@@ -230,46 +235,71 @@ def run_experiments(config):
-            results['experiments'].append({'name': experiment, 'status': 'ERROR', 'description': e_trace})
+        repetition = experiment['repetition']
+        threshold = experiment['threshold']
+        size_a, size_b = get_exp_sizes(experiment)
+
+        for rep in range(repetition):
+            current_experiment = copy.deepcopy(experiment)
+            current_experiment['rep'] = rep + 1
+            logger.info('running experiment: {}'.format(current_experiment))
+            if repetition != 1:
+                logger.info('\trepetition {} out of {}'.format(rep + 1, repetition))
+            result = run_single_experiment(server, config, threshold, size_a, size_b, current_experiment)
+            results['experiments'].append(result)
+
     return results
 
 
+def run_single_experiment(server, config, threshold, size_a, size_b, experiment):
+    result = {}
+    credentials = {}
+    run = {}
+    try:
+        credentials = rest_client.project_create(server, config['schema'], 'mapping',
+                                                 "benchy_{}".format(experiment))
+        # upload clks
+        upload_binary_clks(config, size_a, size_b, credentials)
+        # create run
+        run = rest_client.run_create(server, credentials['project_id'], credentials['result_token'],
+                                     threshold,
+                                     "{}_{}".format(experiment, threshold))
+        # wait for result
+        run_id = run['run_id']
+        result['run_id'] = run_id
+        logger.info(f'waiting for run {run_id} to finish')
+        status = rest_client.wait_for_run(server, credentials['project_id'], run['run_id'],
+                                          credentials['result_token'], timeout=config['timeout'])
+        if status['state'] != 'completed':
+            raise RuntimeError('run did not finish!\n{}'.format(status))
+        logger.info('experiment successful. Evaluating results now...')
+        mapping = rest_client.run_get_result_text(server, credentials['project_id'], run['run_id'],
+                                                  credentials['result_token'])
+        mapping = json.loads(mapping)['mapping']
+        mapping = {int(k): int(v) for k, v in mapping.items()}
+        tt = score_mapping(mapping, *load_truth(config, size_a, size_b))
+        result.update(compose_result(status, tt, experiment, (size_a, size_b), threshold))
+    except Exception as e:
+        e_trace = format_exc()
+        logger.warning("experiment '{}' failed: {}".format(experiment, e_trace))
+        result.update({'name': experiment, 'status': 'ERROR', 'description': e_trace})
+    finally:
+        logger.info('cleaning up...')
+        delete_resources(config, credentials, run)
+
+    return result
+
+
+def push_result_s3(experiment_file):
+    client = boto3.client(
+        's3',
+        aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID'),
```

Comment:
My hunch is we should be agnostic as multiple services offer the S3 api (including MINIO which we already use in this project). 

How about `OBJECT_STORE_SERVER`, `OBJECT_STORE_ACCESS_KEY`,  `OBJECT_STORE_BUCKET` etc as the environment variables? And open an issue to change them in the backend settings from `MINIO_` - https://github.com/data61/anonlink-entity-service/blob/develop/backend/entityservice/settings.py#L27

Reply:
I'm actually not convinced. My reasoning being the following:
 - we do not support anything else than aws in the benchmark
 - the benchmark is fully independent from the main repository, so variable names from the main repo should not impact choices made for the benchmark (even if consistency may be a good idea)
 - but if we are thinking about both repos together, I would actually made them even more specific, such as `AWS_BENCHMARK_...` to be sure that we cannot imagine they will be used for any other context than the chosen one. I would also point out that if we implement a script doing all at once (deploying and benchmarking), it's good to use different environment variable names to be sure that a mis-configuration will not lead to misuse of a token. E.g. I create a script deploying the entity service setting the env var `OBJECT_STORE_ACCESS_KEY`, if the script also starts the benchmark but I forgot to update the env var, I would push the results to the same bucket. Here it is not important as we are not deleting anything but if we were, we could do really bad things...
 - would there be a scenario where a single service has multiple keys? E.g. one key for the bucket `x` and one key for the bucket `y`? The underlying question being if we prefer to have keys per application, or keys per use-case (the application can do x, y and z, or key a can do x, key b can do y and key c can do z, and I give the application the keys a, b and c). Both have pros and cons, but we may want to think about it
- if we generalise to `OBJECT_STORE_SERVER`, I think we will also need an extra field `OBJECT_STORE_TYPE`, in which case we may also need to modify the description of the env variables: one object store type may not use a bucket and access key, but something totally different (I have no clue what it could be, but maybe :) ). This field would also help to know which service to use. 

But the more options we are adding, the more I would assume we should push them into a command line tool instead of env vars.

While not being convinced, I'm also not really strongly opinionated on this, so happy to change if that's the preference :)

Reply:
I'm not too fussed with the variable names but it would be nice to include a way to set the object store server so the benchmark user can decide for themselves.

See the docs at https://docs.min.io/docs/minio-select-api-quickstart-guide.html it looks like all you would need is to add an `endpoint_url` which is only set if an environment variable is present e.g. `S3_SERVER` or my preference `OBJECT_STORE_SERVER`.

```python
s3 = boto3.client('s3',
                  endpoint_url='http://localhost:9000',
                  aws_access_key_id='minio',
                  aws_secret_access_key='minio123',
                  )
```

I don't think we need to tell the benchmark component the `TYPE` of object store - that would defeat the purpose of an abstraction - in this case the S3 API.

Reply:
`OBJECT_STORE_SERVER` added, and other ones renamed to: `OBJECT_STORE_ACCESS_KEY`, `OBJECT_STORE_SECRET_KEY` and `OBJECT_STORE_BUCKET`